[RFC] Python: fixes for sys.path_hooks handler

blueyed opened the pull request:
The main fix is for Python 3.4, which loads a module twice currently (https://github.com/neovim/neovim/issues/2909).

While at it, I've fixed some more issues.

blueyed committed:
python: VimPathFinder: use find_spec for Python 3.4

Fixes https://github.com/neovim/neovim/issues/2909

blueyed committed:
python: VimModuleLoader: check sys.modules[fullname]

This is required by PEP302 for `reload()`ing modules.

blueyed committed:
python: path_hook: properly implement PEP302

The path hook used to load the module already in the `find_module` hook.

This caused different behaviour between Python 2.7 and 3.3, where the
former would call the `VimModuleLoader`, while Python 3.3 appears to
short-circuited this (because the module was loaded already).

This patch will now only find the module, but not load it in the
`find_module` hook.

blueyed committed:
python: VimPathFinder: remove unused load_module method

This method was not used, but `VimModuleLoader.load_module` is.

Oberon00 commented on runtime/autoload/provider/script_host.py:
```diff
@@ -198,22 +204,25 @@ def _find_module(fullname, oldtail, path):
+            return imp.load_module(fullname, *self.module)
 
     class VimPathFinder(object):
         @classmethod
```
This should probably become an [`@staticmethod`](https://docs.python.org/3/library/functions.html#staticmethod), since the `cls` parameter is not used at all. Same with `find_spec`.

blueyed replied:
It uses `self.module`.

blueyed replied:
Done.

justinmk commented:
LGTM. @fwalch any comment?

blueyed commented:
Addressed all issues/comments and rebased on master.

Thanks for the feedback/review!

Marking as `[RFC]` (which I forgot previously).  Should be good to go to `[RDY]` then promptly.

justinmk commented:
@blueyed Thanks
